File: 4th Semester/Systems for Design and Implementation/Java/Homework6/OnwardPersistence/src/main/java/repository/database/UserDbRepository.java
package repository.database;

import models.User;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import repository.UserRepository;
import repository.exceptions.DataChangeException;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import java.util.Properties;
import java.util.UUID;

public class UserDbRepository implements UserRepository {
    private final DbUtils dbUtils;
    private static final Logger logger = LogManager.getLogger(UserDbRepository.class);

    public UserDbRepository(Properties props) {
        logger.traceEntry("Initializing UserDbRepository with properties: {}", props);
        dbUtils = new DbUtils(props);
    }

    protected static User extractUser(ResultSet resultSet) {
        logger.traceEntry("Result set: {}", resultSet);
        try {
            UUID identifier = resultSet.getObject("user_id", UUID.class);
            String username = resultSet.getString("username");
            String password = resultSet.getString("password");
            return logger.traceExit(new User(identifier, username, password));
        } catch (SQLException ex) {
            logger.error(ex);
            throw new RuntimeException(ex);
        }
    }

    @Override
    public UUID save(User entity) throws DataChangeException {
        logger.traceEntry("models.User to save: {}", entity);
        try (PreparedStatement preparedStatement = dbUtils.getConnection().prepareStatement(
                "INSERT INTO users(username, password) VALUES (?, ?)")) {
            preparedStatement.setString(1, entity.getUsername());
            preparedStatement.setString(2, entity.getPassword());
            preparedStatement.executeUpdate();
        } catch (SQLException ex) {
            logger.error(ex);
            throw new DataChangeException(ex);
        }
        try (PreparedStatement preparedStatement = dbUtils.getConnection().prepareStatement(
                "SELECT * FROM users WHERE username = ?")) {
            preparedStatement.setString(1, entity.getUsername());
            var result = DbUtils.findQueryResult(preparedStatement, UserDbRepository::extractUser);
            return logger.traceExit("ID of the new user: {}", result.isEmpty() ? null : result.get(0).getID());
        } catch (SQLException ex) {
            logger.error(ex);
            throw new DataChangeException(ex);
        }
    }

    @Override
    public User find(UUID identifier) {
        logger.traceEntry("ID of the user to find: {}", identifier);
        try (PreparedStatement preparedStatement = dbUtils.getConnection().prepareStatement(
                "SELECT * FROM users WHERE user_id = ?")) {
            preparedStatement.setObject(1, identifier, Types.OTHER);
            var result = DbUtils.findQueryResult(preparedStatement, UserDbRepository::extractUser);
            return logger.traceExit( "Found user: {}", result.isEmpty() ? null : result.get(0));
        } catch (SQLException ex) {
            logger.error(ex);
            throw new RuntimeException(ex);
        }
    }

    @Override
    public User update(User entity) {
        return null;
    }

    @Override
    public User delete(UUID identifier) {
        return null;
    }

    @Override
    public User tryLogin(String username, String password) {
        logger.traceEntry("Username of the user who is trying to log in: {}", username);
        try (PreparedStatement preparedStatement = dbUtils.getConnection().prepareStatement(
                "SELECT * FROM users WHERE username = ? AND password = ?")) {
            preparedStatement.setString(1, username);
            preparedStatement.setString(2, password);
            var result = DbUtils.findQueryResult(preparedStatement, UserDbRepository::extractUser);
            return logger.traceExit("Logged in user: {}", result.isEmpty() ? null : result.get(0));
        } catch (SQLException ex) {
            logger.error(ex);
            throw new RuntimeException(ex);
        }
    }
}

File: 4th Semester/Systems for Design and Implementation/C#/Homework2/OnwardTravel/Repository/Database/UserDbRepository.cs
using log4net;
using log4net.Util;
using OnwardTravel.Models;
using OnwardTravel.Utils;
using System.Data;



namespace OnwardTravel.Repository.Database
{
    internal class UserDbRepository
        : IUserRepository
    {
        private readonly DbUtils DbUtils;
        private static readonly ILog Logger = LogManager.GetLogger("UserDbRepository");

        public UserDbRepository(Properties props)
        {
            DbUtils = new(props);
        }

        protected static User ExtractUser(IDataReader dataR)
        {
            Guid guid = dataR.GetGuid(0);
            string username = dataR.GetString(1);
            string password = dataR.GetString(2);
            return new User(guid, username, password);
        }

        public Guid? Save(User entity)
        {
            Logger.InfoFormat("Saving the user: {0}", entity);
            using (var comm = DbUtils.GetConnection().CreateCommand())
            {
                comm.CommandText = "INSERT INTO users(username, password) VALUES (@username, @password)";
                var paramUsername = comm.CreateParameter();
                paramUsername.ParameterName = "@username";
                paramUsername.Value = entity.Username;
                comm.Parameters.Add(paramUsername);

                var paramPassword = comm.CreateParameter();
                paramPassword.ParameterName = "@password";
                paramPassword.Value = entity.Password;
                comm.Parameters.Add(paramPassword);

                var result = comm.ExecuteNonQuery();
                if (result == 0)
                {
                    Logger.InfoExt(null);
                    return null;
                }
            }
            using (var comm = DbUtils.GetConnection().CreateCommand())
            {
                comm.CommandText = "SELECT user_id FROM users WHERE username = @username";
                var paramUsername = comm.CreateParameter();
                paramUsername.ParameterName = "@username";
                paramUsername.Value = entity.Username;
                comm.Parameters.Add(paramUsername);

                var result = DbUtils.FindQueryResults(comm, DbUtils.ExtractGuid);
                if (result.Count == 0)
                {
                    Logger.InfoExt(null);
                    return null;
                }
                Logger.InfoExt(result[0]);
                return result[0];
            }
        }

        public User? Find(Guid identifier)
        {
            Logger.InfoFormat("Finding the user with the identifier: {0}", identifier);
            
            using (var comm = DbUtils.GetConnection().CreateCommand())
            {
                comm.CommandText = "SELECT * FROM users WHERE user_id = @user_id";
                var paramID = comm.CreateParameter();
                paramID.ParameterName = "@user_id";
                paramID.Value = identifier;
                comm.Parameters.Add(paramID);

                var results = DbUtils.FindQueryResults(comm, ExtractUser);
                if (results.Count == 0)
                {
                    Logger.InfoExt(null);
                    return null;
                }
                var user = results[0];
                Logger.InfoExt(user);
                return user;
            }
        }

        public User? Update(User entity)
        {
            Logger.InfoFormat("Updating the user with the identifier: {0}", entity.ID);

            User? oldUser = Find(entity.ID);
            if (oldUser == null)
            {
                Logger.InfoExt(null);
                return null;
            }

            using (var comm = DbUtils.GetConnection().CreateCommand())
            {
                comm.CommandText = "UPDATE users SET username = @username, password = @password WHERE user_id = @id";
                var paramUsername = comm.CreateParameter();
                paramUsername.ParameterName = "@username";
                paramUsername.Value = entity.Username;
                comm.Parameters.Add(paramUsername);

                var paramPassword = comm.CreateParameter();
                paramPassword.ParameterName = "@password";
                paramPassword.Value = entity.Password;
                comm.Parameters.Add(paramPassword);

                var paramID = comm.CreateParameter();
                paramID.ParameterName = "@id";
                paramID.Value = entity.ID;
                comm.Parameters.Add(paramID);

                var result = comm.ExecuteNonQuery();
                if (result == 0)
                {
                    Logger.InfoExt(null);
                    return null;
                }
            }

            Logger.InfoExt(oldUser);
            return oldUser;
        }

        public User? Delete(Guid identifier)
        {
            Logger.InfoFormat("Deleting the user with the identifier: {0}", identifier);

            User? oldUser = Find(identifier);
            if (oldUser == null)
            {
                Logger.InfoExt(null);
                return null;
            }

            using (var comm = DbUtils.GetConnection().CreateCommand())
            {
                comm.CommandText = "DELETE FROM users WHERE user_id = @id";
                var paramID = comm.CreateParameter();
                paramID.ParameterName = "@id";
                paramID.Value = identifier;
                comm.Parameters.Add(paramID);

                var result = comm.ExecuteNonQuery();
                if (result == 0)
                {
                    Logger.InfoExt(null);
                    return null;
                }
            }

            Logger.InfoExt(oldUser);
            return oldUser;
        }

        public User? TryLogin(string username, string password)
        {
            Logger.InfoFormat("Trying to login the user with the username: {0}", username);

            using (var comm = DbUtils.GetConnection().CreateCommand())
            {
                comm.CommandText = "SELECT * FROM users WHERE username = @username AND password = @password";
                var paramUsername = comm.CreateParameter();
                paramUsername.ParameterName = "@username";
                paramUsername.Value = username;
                comm.Parameters.Add(paramUsername);

                var paramPassword = comm.CreateParameter();
                paramPassword.ParameterName = "@password";
                paramPassword.Value = password;
                comm.Parameters.Add(paramPassword);

                var results = DbUtils.FindQueryResults(comm, ExtractUser);
                if (results.Count == 0)
                {
                    Logger.InfoExt(null);
                    return null;
                }
                var user = results[0];
                Logger.InfoExt(user);
                return user;
            }
        }
    }
}
